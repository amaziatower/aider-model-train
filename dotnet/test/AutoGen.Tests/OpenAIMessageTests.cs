// Copyright (c) Microsoft Corporation. All rights reserved.
// OpenAIMessageTests.cs

using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using ApprovalTests;
using ApprovalTests.Namers;
using ApprovalTests.Reporters;
using AutoGen.OpenAI;
using Azure.AI.OpenAI;
using Xunit;

namespace AutoGen.Tests;

public class OpenAIMessageTests
{
    private readonly JsonSerializerOptions jsonSerializerOptions = new JsonSerializerOptions
    {
        WriteIndented = true,
        IgnoreReadOnlyProperties = false,
    };

    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("ApprovalTests")]
    public void BasicMessageTest()
    {
        IMessage[] messages = [
            new TextMessage(Role.System, "You are a helpful AI assistant"),
            new TextMessage(Role.User, "Hello", "user"),
            new TextMessage(Role.Assistant, "How can I help you?", from: "assistant"),
            new Message(Role.System, "You are a helpful AI assistant"),
            new Message(Role.User, "Hello", "user"),
            new Message(Role.Assistant, "How can I help you?", from: "assistant"),
            new Message(Role.Function, "result", "user"),
            new Message(Role.Assistant, null, "assistant")
            {
                FunctionName = "functionName",
                FunctionArguments = "functionArguments",
            },
            new ImageMessage(Role.User, "https://example.com/image.png", "user"),
            new MultiModalMessage(
                [
                    new TextMessage(Role.User, "Hello", "user"),
                    new ImageMessage(Role.User, "https://example.com/image.png", "user"),
                ], "user"),
            new ToolCallMessage("test", "test", "assistant"),
            new ToolCallResultMessage("result", new ToolCallMessage("test", "test", "assistant"), "user"),
            new ParallelToolCallResultMessage(
                [
                    new ToolCallResultMessage("result", new ToolCallMessage("test", "test", "assistant"), "user"),
                    new ToolCallResultMessage("result", new ToolCallMessage("test", "test", "assistant"), "user"),
                ], "user"),
            new AggregateMessage(
                [
                    new ToolCallMessage("test", "test", "assistant"),
                    new ToolCallMessage("test", "test", "assistant"),
                ], "assistant"),
        ];

        var agent = new EchoAgent("assistant");

        var oaiMessages = messages.Select(m => (m, agent.ToOpenAIChatRequestMessage(m).Select(m => m.Content)));
        VerifyOAIMessages(oaiMessages);
    }

    private void VerifyOAIMessages(IEnumerable<(IMessage, IEnumerable<ChatRequestMessage>)> messages)
    {
        var jsonObjects = messages.Select(pair =>
        {
            var (originalMessage, ms) = pair;
            var objs = new List<object>();
            foreach (var m in ms)
            {
                object? obj = null;
                if (m is ChatRequestUserMessage userMessage)
                {
                    obj = new
                    {
                        Role = userMessage.Role.ToString(),
                        Content = userMessage.Content,
                        MultiModaItem = userMessage.MultimodalContentItems?.Select(item =>
                        {
                            return item switch
                            {
                                ChatMessageImageContentItem imageContentItem => new
                                {
                                    Type = "Image",
                                    ImageUrl = imageContentItem.ImageUrl,
                                } as object,
                                ChatMessageTextContentItem textContentItem => new
                                {
                                    Type = "Text",
                                    Text = textContentItem.Text,
                                } as object,
                                _ => throw new System.NotImplementedException(),
                            };
                        }),
                    };
                }

                if (m is ChatRequestAssistantMessage assistantMessage)
                {
                    obj = new
                    {
                        Role = assistantMessage.Role.ToString(),
                        Content = assistantMessage.Content,
                        TooCall = assistantMessage.ToolCalls.Select(tc =>
                        {
                            return tc switch
                            {
                                ChatCompletionsFunctionToolCall functionToolCall => new
                                {
                                    Type = "Function",
                                    Name = functionToolCall.Name,
                                    Arguments = functionToolCall.Arguments,
                                    Id = functionToolCall.Id,
                                } as object,
                                _ => throw new System.NotImplementedException(),
                            };
                        }),
                        FunctionCallName = assistantMessage.FunctionCall?.Name,
                        FunctionCallArguments = assistantMessage.FunctionCall?.Arguments,
                    };
                }

                if (m is ChatRequestSystemMessage systemMessage)
                {
                    obj = new
                    {
                        Role = systemMessage.Role.ToString(),
                        Content = systemMessage.Content,
                    };
                }

                if (m is ChatRequestFunctionMessage functionMessage)
                {
                    obj = new
                    {
                        Role = functionMessage.Role.ToString(),
                        Content = functionMessage.Content,
                        Name = functionMessage.Name,
                    };
                }

                if (m is ChatRequestToolMessage toolCallMessage)
                {
                    obj = new
                    {
                        Role = toolCallMessage.Role.ToString(),
                        Content = toolCallMessage.Content,
                        ToolCallId = toolCallMessage.ToolCallId,
                    };
                }

                objs.Add(obj ?? throw new System.NotImplementedException());
            }

            return new
            {
                OriginalMessage = originalMessage.ToString(),
                ConvertedMessages = objs,
            };
        });

        var json = JsonSerializer.Serialize(jsonObjects, this.jsonSerializerOptions);
        Approvals.Verify(json);
    }
}
