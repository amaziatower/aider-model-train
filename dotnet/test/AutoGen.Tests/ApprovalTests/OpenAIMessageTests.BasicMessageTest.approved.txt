[
  {
    "OriginalMessage": "TextMessage(system, You are a helpful AI assistant, )",
    "ConvertedMessages": [
      {
        "Role": "system",
        "Content": "You are a helpful AI assistant"
      }
    ]
  },
  {
    "OriginalMessage": "TextMessage(user, Hello, user)",
    "ConvertedMessages": [
      {
        "Role": "user",
        "Content": "Hello",
        "MultiModaItem": null
      }
    ]
  },
  {
    "OriginalMessage": "TextMessage(assistant, How can I help you?, assistant)",
    "ConvertedMessages": [
      {
        "Role": "assistant",
        "Content": "How can I help you?",
        "TooCall": [],
        "FunctionCallName": null,
        "FunctionCallArguments": null
      }
    ]
  },
  {
    "OriginalMessage": "Message(system, You are a helpful AI assistant, , , )",
    "ConvertedMessages": [
      {
        "Role": "system",
        "Content": "You are a helpful AI assistant"
      }
    ]
  },
  {
    "OriginalMessage": "Message(user, Hello, user, , )",
    "ConvertedMessages": [
      {
        "Role": "user",
        "Content": "Hello",
        "MultiModaItem": null
      }
    ]
  },
  {
    "OriginalMessage": "Message(assistant, How can I help you?, assistant, , )",
    "ConvertedMessages": [
      {
        "Role": "assistant",
        "Content": "How can I help you?",
        "TooCall": [],
        "FunctionCallName": null,
        "FunctionCallArguments": null
      }
    ]
  },
  {
    "OriginalMessage": "Message(function, result, user, , )",
    "ConvertedMessages": [
      {
        "Role": "user",
        "Content": "result",
        "MultiModaItem": null
      }
    ]
  },
  {
    "OriginalMessage": "Message(assistant, , assistant, functionName, functionArguments)",
    "ConvertedMessages": [
      {
        "Role": "assistant",
        "Content": null,
        "TooCall": [],
        "FunctionCallName": "functionName",
        "FunctionCallArguments": "functionArguments"
      }
    ]
  },
  {
    "OriginalMessage": "ImageMessage(user, https://example.com/image.png, user)",
    "ConvertedMessages": [
      {
        "Role": "user",
        "Content": null,
        "MultiModaItem": [
          {
            "Type": "Image",
            "ImageUrl": {
              "Url": "https://example.com/image.png",
              "Detail": null
            }
          }
        ]
      }
    ]
  },
  {
    "OriginalMessage": "MultiModalMessage(, user)\n\tTextMessage(user, Hello, user)\n\tImageMessage(user, https://example.com/image.png, user)",
    "ConvertedMessages": [
      {
        "Role": "user",
        "Content": null,
        "MultiModaItem": [
          {
            "Type": "Text",
            "Text": "Hello"
          },
          {
            "Type": "Image",
            "ImageUrl": {
              "Url": "https://example.com/image.png",
              "Detail": null
            }
          }
        ]
      }
    ]
  },
  {
    "OriginalMessage": "ToolCallMessage(test, test, assistant)",
    "ConvertedMessages": [
      {
        "Role": "assistant",
        "Content": "",
        "TooCall": [
          {
            "Type": "Function",
            "Name": "test",
            "Arguments": "test",
            "Id": "test"
          }
        ],
        "FunctionCallName": null,
        "FunctionCallArguments": null
      }
    ]
  },
  {
    "OriginalMessage": "ToolCallResultMessage(result, ToolCallMessage(test, test, assistant), user)",
    "ConvertedMessages": [
      {
        "Role": "tool",
        "Content": "result",
        "ToolCallId": "test"
      }
    ]
  },
  {
    "OriginalMessage": "ParallelToolCallResultMessage(user)\n\tToolCallResultMessage(result, ToolCallMessage(test, test, assistant), user)\n\tToolCallResultMessage(result, ToolCallMessage(test, test, assistant), user)",
    "ConvertedMessages": [
      {
        "Role": "tool",
        "Content": "result",
        "ToolCallId": "test"
      },
      {
        "Role": "tool",
        "Content": "result",
        "ToolCallId": "test"
      }
    ]
  },
  {
    "OriginalMessage": "AggregateMessage(assistant)\n\tToolCallMessage(test, test, assistant)\n\tToolCallMessage(test, test, assistant)",
    "ConvertedMessages": [
      {
        "Role": "assistant",
        "Content": "",
        "TooCall": [
          {
            "Type": "Function",
            "Name": "test",
            "Arguments": "test",
            "Id": "test"
          },
          {
            "Type": "Function",
            "Name": "test",
            "Arguments": "test",
            "Id": "test"
          }
        ],
        "FunctionCallName": null,
        "FunctionCallArguments": null
      }
    ]
  }
]